ipa-core/src/protocol/hybrid/agg.rs

```diff
@@ -0,0 +1,235 @@
+/// It groups all the reports by the PRF of the `match_key`, finds all reports from `match_keys`
+/// with that provided exactly 2 reports, then adds those 2 reports.
+pub async fn aggregate_reports<BK, V, C>(
+    ctx: C,
+    reports: Vec<PrfHybridReport<BK, V>>,
+) -> Result<Vec<AggregateableHybridReport<BK, V>>, Error>
+where
+    C: UpgradableContext,
+    BK: SharedValue + BooleanArray,
+    V: SharedValue + BooleanArray,
+    Replicated<Boolean>: BooleanProtocols<DZKPUpgraded<C>>,
+{
+    let mut reports_by_matchkey = HashMap::with_capacity(reports.len() / 2);
+
+    // build a hashmap of match_key -> ([AggregateableHybridReport;2], count)
+    // if count ever exceeds 2, we drop reports, but keep counting
+    // an honest client and report collector will only submit
+    // one report with a breakdown key and one report with a value.
+    // if there are less, it's unattributed. if more, something went wrong.
+    for report in reports {
+        let match_key = report.match_key;
+        let entry = reports_by_matchkey.entry(match_key).or_insert((
+            [
+                AggregateableHybridReport::<BK, V>::ZERO,
+                AggregateableHybridReport::<BK, V>::ZERO,
+            ],
+            0,
+        ));
+        if entry.1 == 0 {
+            // If the count is 0, replace the first element with the new report
+            entry.0[0] = report.into();
+            entry.1 += 1;
+        } else if entry.1 == 1 {
+            // If the count is 1, replace the second element with the new report
+            entry.0[1] = report.into();
+            entry.1 += 1;
+        } else {
+            // If the count is 2 or more, increment the counter and drop the report
+            entry.1 += 1;
+        }
+    }
+
+    // we only keep the reports from match_keys that provided exactly 2 reports
+    let report_pairs: Vec<[AggregateableHybridReport<BK, V>; 2]> = reports_by_matchkey
+        .into_iter()
+        .filter_map(|(_, v)| if v.1 == 2 { Some(v.0) } else { None })
+        .collect::<Vec<_>>();
+
+    // let mut agg_reports = Vec::with_capacity(report_pairs.len());
+
+    let chunk_size = TARGET_PROOF_SIZE;
+
+    let dzkp_validator = ctx.clone().dzkp_validator(
+        MaliciousProtocolSteps {
+            protocol: &HybridStep::GroupBySum,
+            validate: &HybridStep::GroupBySumValidate,
+        },
+        std::cmp::min(ctx.active_work().get(), chunk_size.next_power_of_two()),
+    );
+
+    let ctx = dzkp_validator
+        .context()
+        .set_total_records(TotalRecords::specified(2 * report_pairs.len())?);
+
+    let agg_work = stream::iter(report_pairs)
+        .enumerate()
+        .map(|(idx, reports)| {
+            let record_id_bk = RecordId::FIRST + 2 * idx;
+            let record_id_v = RecordId::FIRST + 2 * idx + 1;
+            let agg_ctx = ctx.clone();
+            async move {
+                let (breakdown_key, _) = integer_add::<_, EightBitStep, 1>(
```

Comment:
how do we pick the step size here? does it depend on the size of BK and V? if so, how do we do it with generics?

Reply:
If we don't have enough bits in step, it will panic. Using a step that is too large may lead to compact step size blowout, but it may not be a concern for hybrid. We can probably pick a reasonable number of bits for both (whatever we use in the aggregation protocol)

Reply:
EightBit is probably sufficient for both of these for now, then. We'll need a larger one once we aggregate multiple values.

Reply:
iirc for breakdown keys the additions don't matter as it is a sum with 0. for trigger values we may need to handle overflow

Reply:
at this point, we're assuming that there is exactly 1 conversion and 1 impression, so the same is true for the value here. not true when we have multiple values to aggregate though (later in the protocol)